src/Common/src/System/Security/Cryptography/Asn1V2.Serializer.cs

```diff
@@ -0,0 +1,1431 @@
+                // type modification for any of the primitive types.  If that changes, we
+                // need to either pass it through here or do some other form of rerouting.
+                Debug.Assert(!fieldData.WasCustomized);
+
+                return GetPrimitiveDeserializer(typeT, expectedTag);
+            }
+
+            if (typeT.IsEnum)
+            {
+                if (typeT.GetCustomAttributes(typeof(FlagsAttribute), false).Length > 0)
+                {
+                    return reader => reader.GetNamedBitListValue(expectedTag, typeT);
+                }
+
+                return reader => reader.GetEnumeratedValue(expectedTag, typeT);
+            }
+
+            if (typeT == typeof(string))
+            {
+                if (fieldData.TagType == UniversalTagNumber.ObjectIdentifier)
+                {
+                    return reader => reader.ReadObjectIdentifierAsString(expectedTag);
+                }
+
+                return reader => reader.GetCharacterString(expectedTag, localFieldData.TagType);
+            }
+
+            if (typeT == typeof(ReadOnlyMemory<byte>) && !fieldData.IsCollection)
+            {
+                if (fieldData.IsAny)
+                {
+                    Asn1Tag matchTag = fieldData.ExpectedTag;
+                    
+                    // EXPLICIT Any can't be validated, just return the value.
+                    // And if no field type was known, just read that, too.
+                    if (fieldData.HasExplicitTag || matchTag == Asn1Tag.EndOfContents)
+                    {
+                        return reader => reader.GetEncodedValue();
+                    }
+
+                    return reader =>
+                    {
+                        Asn1Tag nextTag = reader.PeekTag();
+
+                        if (matchTag.TagClass != nextTag.TagClass ||
+                            matchTag.TagValue != nextTag.TagValue)
+                        {
+                            throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                        }
+
+                        return reader.GetEncodedValue();
+                    };
+                }
+
+                if (fieldData.TagType == UniversalTagNumber.BitString)
+                {
+                    return reader =>
+                    {
+                        if (reader.TryGetPrimitiveBitStringValue(expectedTag, out _, out ReadOnlyMemory<byte> contents))
+                        {
+                            return contents;
+                        }
+
+                        // Guaranteed too big, because it has the tag and length.
+                        int length = reader.PeekEncodedValue().Length;
+                        byte[] rented = ArrayPool<byte>.Shared.Rent(length);
+                        
+                        try
+                        {
+                            if (reader.TryCopyBitStringBytes(rented, out _, out int bytesWritten))
```

Comment:
Do we need a way of communicating to the deserialized object how many leftover bits there are? I assume a POCO author would only use bitstring if he's expecting unused bits (otherwise he'd use octet string), so the leftover bit count seems like data he'd care about.

Reply:
For now that seems like a "sorry, you have to use the reader" scenario.  Perhaps a (Value)Tuple of int and "a thing that holds data" could be used in the future.